File: lab/mergeSort.js
function merge(arr1, arr2) {
	let [i,j] = [0,0];

	let result = [];
	while (i < arr1.length || j < arr2.length) {

		// First check to see if i and j are in bounds 
		if (i >= arr1.length) {
			if (j < arr2.length) {
				return [...result, ...arr2.slice(j)]
			}
		}

		if (j >= arr2.length) {
			if (i < arr1.length) {
				return [...result, ...arr1.slice(i)]
			}
		}
	

		if (arr1[i] > arr2[j]) {
			result.push(arr2[j]);
			j++;
		} else if (arr1[j] === arr2[i]) {
			result.push(arr1[j], arr2[i]); 
			i++;
			j++;
		} else {
			result.push(arr1[i]);
			i++;
		}
	}
	return result;
}

function mergeSort(arr) {
	
	if (arr.length === 1) {
		return arr;
	}
	
	let q = Math.floor(arr.length/2);
	let l = arr.slice(0,q);
	let r = arr.slice(q);

	console.log(l, r) 
	return merge(mergeSort(l), mergeSort(r))
}

console.log(mergeSort([1,5,2,7]))

File: lab/merge_sort.py
def merge(a, b): 
    i = 0
    j = 0
    sorted = []
    inversions = a[1] + b[1]
    a = a[0]
    b = b[0]
    while (i < len(a) and j < len(b)): 
        
        if a[i] < b[j]: 
            sorted.append(a[i])
            i+=1
        elif a[i] == b[j]: 
            sorted.append(a[i])
            sorted.append(b[j])
            i += 1
            j += 1
        else:
            inversions += 1 
            sorted.append(b[j])
            j+=1            

    if i <= len(a) - 1: 
        sorted = sorted + a[i:] 
    elif j <= len(b) - 1: 
        sorted = sorted + b[j:] 

    return [sorted, inversions]

def merge_sort(arr, left, right): 
    if left == right-1: 
        return [arr[left:right], 0]
    
    midpoint = (right-left)//2 + left 

    return merge(merge_sort(arr, left, midpoint), merge_sort(arr, midpoint, right))
